### PR TITLE
Add csrf token support

### PR DESCRIPTION
**Pull request by olOwOlo**
#### What's this PR do/fix?

Try to find csrf token and add it to the header of all requests by patching the `requestInterceptor`. The behavior is based on the configuration described in the spring security reference. The following order is the order of finding the csrf token.

P.S. This implementation will always use the csrf token obtained the first time.

P.S. If necessary, Maybe someone in the future can add the ability to configure these behaviors, I don’t know how to add options. 😢 

1. **CSRF token within your meta tags** [<sup>[1]</sup>](https://docs.spring.io/spring-security/site/docs/5.0.5.RELEASE/reference/htmlsingle/#csrf-include-csrf-token-ajax)

    The meta name must be `_csrf` and `_csrf_header`. `/` is the page used to query meta data.

    ```html
    <html>
    <head>
      <meta name="_csrf" content="${_csrf.token}"/>
      <!-- default header name is X-CSRF-TOKEN -->
      <meta name="_csrf_header" content="${_csrf.headerName}"/>
    </head>
    ```

2. **Define an endpoint used to fetch csrf token** [<sup>[2]</sup>](https://docs.spring.io/spring-security/site/docs/5.0.5.RELEASE/reference/htmlsingle/#mvc-csrf-resolver)

    The `@RequestMapping` value must be `/csrf`.

    ```java
    @RestController
    public class CsrfController {
    
      @RequestMapping("/csrf")
      public CsrfToken csrf(CsrfToken token) {
        return token;
      }
    }
    ```

    > It is important to keep the `CsrfToken` a secret from other domains.
This means if you are using [Cross Origin Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS), you should **NOT** expose the `CsrfToken` to any external domains.

3. **CSRF token within your cookie** [<sup>[3]</sup>](https://docs.spring.io/spring-security/site/docs/5.0.5.RELEASE/reference/htmlsingle/#csrf-cookie)

    The default cookie name is `XSRF-TOKEN`, and the default header name is `X-XSRF-TOKEN`.

    ```java
    @EnableWebSecurity
    public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
    
    	@Override
    	protected void configure(HttpSecurity http) throws Exception {
    		http.csrf().csrfTokenRepository(CookieCsrfTokenRepository.withHttpOnlyFalse());
    	}
    }
    ```

#### Are there unit tests? If not how should this be manually tested?

File      |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
----------|----------|----------|----------|----------|-------------------|
 csrf.js  |    82.86 |     87.5 |     87.5 |    81.82 | 10,11,13,17,19,21 |

#### Any background context you want to provide?

| Parameter Name | Description |
|:--:|:--:|
| [`requestInterceptor`](https://github.com/swagger-api/swagger-ui/blob/master/docs/usage/configuration.md#network) | `Function=(a => a)`. MUST be a function.  Function to intercept remote definition, Try-It-Out, and OAuth2 requests.  Accepts one argument requestInterceptor(request) and must return the modified request, or a Promise that resolves to the modified request. |

#### What are the relevant issues?

#1450

### CODE REVIEW

**Comment by olOwOlo** [comment]
Hi @dilipkrish , any suggestion?

File: springfox-swagger-ui/src/web/.babelrc [hunk 1]
```diff
@@ -4,6 +4,7 @@
       "targets": {
         "browsers": ["last 2 versions"]
       }
-    }]
+    }],
+    "stage-3"
```
Comment (dilipkrish) [inline springfox-swagger-ui/src/web/.babelrc]: Why do we need this?

Comment (kasecato) [reply springfox-swagger-ui/src/web/.babelrc]: @dilipkrish I think they wants to use [object-rest-spread](https://github.com/springfox/springfox/pull/2434/files#diff-9c74f68d3c9c52c2ea4647ddfb8a3183R14). 
```bash
...request.headers,
```
- [Babel > stage-3 > transform-object-rest-spread](https://babeljs.io/docs/en/babel-preset-stage-3.html)

Comment (olOwOlo) [reply springfox-swagger-ui/src/web/.babelrc]: As @kasecato said.

I will push a new commit to refactor it. If you prefer this way, I will revert the commit.

**Comment by dilipkrish** [comment]
@olOwOlo didnt mean for u to change it, was just asking if that has an impact on what browsers are supported. Thank you for doing that tho'.

Also one last question. Does the swagger-ui.html need to change? Seems like it should have the meta tags as u suggested?

```html
<!-- The meta name must be _csrf and _csrf_header. / is the page used to query meta data. -->

<html>
<head>
  <meta name="_csrf" content="${_csrf.token}"/>
  <!-- default header name is X-CSRF-TOKEN -->
  <meta name="_csrf_header" content="${_csrf.headerName}"/>
</head>
```

**Comment by olOwOlo** [comment]
@dilipkrish No, `swagger-ui.html` is a fully static file. Meta tags should be included in templates, such as JSP, freemarker, thymeleaf, etc.

**Comment by tangxiucai2** [comment]
> 1. CSRF

where is CsrfToken class ??

**Comment by ahoehma** [comment]
Hello guys, sorry but I have to ask :) Is there somewhere a "complete" example how to get rid of "404 for csrf" in spring boot 2.X?

**Comment by liuweiGL** [comment]
This is a terrible pr.

**Comment by bmistry13** [comment]
If CSRF token in a cookie, how can swagger UI pick it up?  The token is changing per ANY POST/PUT/DELETE/PATCH operation.  So, is there any logic to pick up the value from the cookie?

**Pull request merged**